crates/kitsune_p2p/kitsune_p2p/tests/sharding.rs

```diff
@@ -0,0 +1,424 @@
+    let basis = Arc::new(kitsune_basis);
+
+    // If the location was copied correctly then the basis location should be the same as the sender
+    // location. Due to the logic above, the receiver should have the sender's location in its arc.
+    assert_eq!(agents[sender_idx].2.get_loc(), basis.get_loc());
+
+    let test_op = TestHostOp::new(space.clone());
+
+    agents[sender_idx]
+        .0
+        .op_store()
+        .write()
+        .push(test_op.clone());
+
+    agents[sender_idx]
+        .1
+        .broadcast(
+            space.clone(),
+            basis.clone(),
+            KitsuneTimeout::from_millis(5_000),
+            BroadcastData::Publish {
+                source: agents[sender_idx].2.clone(),
+                op_hash_list: vec![test_op.into()],
+                context: FetchContext::default(),
+            },
+        )
+        .await
+        .unwrap();
+
+    tokio::time::timeout(std::time::Duration::from_secs(60), {
+        let op_store_recv = agents[should_recv_idx].0.op_store().clone();
+        async move {
+            loop {
+                if !op_store_recv.read().is_empty() {
+                    break;
+                }
+                tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+            }
+        }
+    })
+    .await
+    .unwrap();
+
+    assert_eq!(1, agents[should_recv_idx].0.op_store().read().len());
+
+    let mut agents_not_receiving_count = 0;
+    for i in 0..5 {
+        if i == sender_idx || i == should_recv_idx {
+            continue;
+        }
+
+        // We've filtered out the sender and the receiver, who are expected to have the data.
+        // Now we check that the agent at the current index does not have the basis that the op was
+        // published to in its arc. That would make the test wrong, not Kitsune, so fail here!
+        let should_this_agent_hold_the_op =
+            should_agent_hold_op_at_basis(&agents[i].0, agents[i].2.clone(), basis.clone());
+
+        if should_this_agent_hold_the_op {
+            tracing::warn!(
+                "Agent {} should not have received the data, this means that the test data construction isn't perfectly accurate but the test can continue",
+                i
+            );
+
+            continue;
+        }
+
+        // Now make the important assertion that the agent at index `i` did not receive the data! If it's not in the agents arc
+        // (which we just asserted above) then it should not have been received.
+        let store_lock = agents[i].0.op_store();
+        let store = store_lock.read();
+        assert!(
+            store.is_empty(),
+            "Agent {} should not have received any data but has {} ops. Ops store: {:?}",
+            i,
+            store.len(),
+            store,
+        );
+        agents_not_receiving_count += 1;
+    }
+
+    // A slightly looser assertion than requiring the test data to be constructed perfectly with no extra overlap.
```

Comment:
Extra overlap being an agent whose arc overlaps with the next two agents instead of just the next one?

Could it be worth actually constructing the test data perfectly by just putting the whole agent selection process in an outer loop, throwing it all out and trying again if extra overlap is detected?

Reply:
Possibly yes, but it's a little tricky to do because I either have to
- Make sure this code converts the raw arc the same way that Kitsune does to check it, otherwise the test will end up flaky because of a mismatch.
- Join the agents and ask Kitsune what the arcs are. If there's an overlap then throw away the Kitsune instances and the bootstrap server so ensure that old agents hanging around don't mess up the test.

So I agree that ideal test data is preferable but it's either a bit of risk with arc conversions or very resource heavy.  I think I prefer the risk and actually and then I could assert that the arcs the test thought the agents would use are identical to the arcs that Kitsune ends up putting on the AgentInfoSigned.

I'll take a look at this and see if I can make it work like that

Reply:
Oh this is less difficult than I thought, Kitsune doesn't do anything with the value I provide. As long as it gets communicated accurately through an 'arc size' then this is simple

Reply:
Done, and I've been able to put the stronger assertions back and the tests are still stable. Thanks for the nudge in this direction! I thought it'd be more complicated than it was :)